substra/sdk/backends/local/backend.py

```diff
@@ -0,0 +1,523 @@
+            models.CompositeAlgo, spec, exist_ok, spec_options=spec_options
+        )
+
+    def __intersect_permissions(self, aggregate_set, element):
+        if not element.permissions.process.public:
+            if aggregate_set is None:
+                aggregate_set = set(element.permissions.process.authorized_ids)
+            else:
+                aggregate_set = aggregate_set & set(
+                    element.permissions.process.authorized_ids
+                )
+        return aggregate_set
+
+    def __add_compute_plan(
+        self,
+        traintuple_keys=None,
+        composite_traintuple_keys=None,
+        aggregatetuple_keys=None,
+        testtuple_keys=None,
+    ):
+        tuple_count = functools.reduce(
+            lambda x, y: x + (len(y) if y else 0),
+            [
+                traintuple_keys,
+                composite_traintuple_keys,
+                aggregatetuple_keys,
+                testtuple_keys,
+            ],
+            0,
+        )
+        compute_plan = models.ComputePlan(
+            compute_plan_id=uuid.uuid4().hex,
+            status=models.Status.waiting.value,
+            traintuple_keys=traintuple_keys,
+            composite_traintuple_keys=composite_traintuple_keys,
+            aggregatetuple_keys=aggregatetuple_keys,
+            testtuple_keys=testtuple_keys,
+            id_to_key=dict(),
+            tag="",
+            tuple_count=tuple_count,
+            done_count=0,
+            metadata=dict(),
+        )
+        return self._db.add(compute_plan)
+
+    def _add_traintuple(self, spec, exist_ok, spec_options=None):
+        # validation
+        algo = self._db.get(schemas.Type.Algo, spec.algo_key)
+        data_manager = self._db.get(schemas.Type.Dataset, spec.data_manager_key)
+        in_traintuples = (
+            [self._db.get(schemas.Type.Traintuple, key) for key in spec.in_models_keys]
+            if spec.in_models_keys is not None
+            else []
+        )
+        same_data_manager = all(
+            [
+                spec.data_manager_key
+                in self._db.get(schemas.Type.DataSample, key).data_manager_keys
+                for key in spec.train_data_sample_keys
+            ]
+        )
+        if not same_data_manager:
+            raise substra.exceptions.InvalidRequest(
+                "dataSample do not belong to the same dataManager", 400
+            )
+
+        key_components = (
+            [_BACKEND_ID, spec.algo_key, spec.data_manager_key]
+            + spec.train_data_sample_keys
+            + spec.in_models_keys
+            if spec.in_models_keys is not None
+            else []
+        )
+        key = hasher.Hasher(values=key_components).compute()
+
+        # permissions
+        with_permissions = [algo, data_manager] + in_traintuples
+
+        authorized_ids = functools.reduce(
+            self.__intersect_permissions, with_permissions, None
+        )
+        public = authorized_ids is None
```

Comment:
What happens if there was an error and for example the algo is private to only A and the dataset is private to only B? Normally the traintuple won't be accepted, but I read the code as "the traintuple is accepted and the model permission will be public".

Reply:
Indeed there is a problem in the case "private and no permission"